ic-helpers/src/ledger/principal_ext.rs

```diff
@@ -37,90 +29,119 @@ pub struct Block{
-        }
+            if AccountIdentifier::from_principal(*self, None) != block.to {
+                println!("{:?}", block.to);
+                return Err(TxError::ErrorTo);
+            }
+
+            if block.amount < threshold {
+                return Err(TxError::AmountTooSmall);
+            }
+
+        Ok(())
+    }
 
-        if block.amount < threshold {
-            return Err(TxError::AmountTooSmall);
+    async fn ledger_transfer(&self, to: Principal,  amount: u64, fee: u64) -> Result<u64, String> {
+        if amount < fee {
+            return Err(format!("Cannot transfer tokens. Amount `{}` is smaller then the fee `{}`.", amount, fee));
         }
 
-    Ok(())
-}
+        let args = SendArgs {
+            memo: Memo(0x57444857),
+            amount: (Tokens::from_e8s(amount) - Tokens::from_e8s(fee))?,
+            fee: Tokens::from_e8s(fee),
+            from_subaccount: None,
+            to: AccountIdentifier::from_principal(to, None),
+            created_at_time: None,
+        };
+        let result: Result<(u64,),_> = ic_cdk::call(*self,"send_dfx",(args,),).await;
+        result
+         .map(|v| v.0)
+         .map_err(|e| format!("{:?}", e))
+    }
 
-async fn ledger_transfer(&self, to: Principal,  amount: u64, fee: u64) -> Result<u64, String> {
-    if amount < fee {
-        return Err(format!("Cannot transfer tokens. Amount `{}` is smaller then the fee `{}`.", amount, fee));
+    async fn get_balance(&self, of: Principal, sub_account: Option<Subaccount>) -> Result<u64, String> {
+        let account = AccountIdentifier::new(of.into(), sub_account);
+        let args = BinaryAccountBalanceArgs { account: account.to_address() };
+        let result = ic_cdk::call::<_, (Tokens,)>(*self, "account_balance", (args,)).await.map_err(|e| e.1)?.0;
+        Ok(result.get_e8s())
     }
-    
-    let args = SendArgs {
-        memo: Memo(0x57444857),
-        amount: (Tokens::from_e8s(amount) - Tokens::from_e8s(fee))?,
-        fee: Tokens::from_e8s(fee),
-        from_subaccount: None,
-        to: AccountIdentifier::from_principal(to, None),
-        created_at_time: None,
-    };
-    let result: Result<(u64,),_> = ic_cdk::call(*self,"send_dfx",(args,),).await;
-    result
-     .map(|v| v.0)
-     .map_err(|e| format!("{:?}", e))
 
+    async fn transfer(&self, to: Principal, amount: u64, from_subaccount: Option<Subaccount>, to_subaccount: Option<Subaccount>) -> Result<u64, String> {
+        if amount < TRANSACTION_FEE.get_e8s() {
+            return Err(format!("cannot transfer tokens: amount '{}' is smaller then the fee '{}'", amount, TRANSACTION_FEE.get_e8s()))
+        }
+
+        let args = TransferArgs {
+            memo: Default::default(),
+            amount: (Tokens::from_e8s(amount) - TRANSACTION_FEE)?,
+            fee: TRANSACTION_FEE,
+            from_subaccount,
+            to: AccountIdentifier::from_principal(to, to_subaccount).to_address(),
+            created_at_time: None
```

Comment:
Should this be set?

Reply:
Doesn't need to be set. IC sets it to the current TS for the caller anyway.